controllers/provisionserver_controller.go

```diff
@@ -130,37 +137,46 @@ func (r *ProvisionServerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
 		return ctrl.Result{}, err
 	}
 
-	// provisionserver
-	// Create or update the Deployment object
-	op, err := r.deploymentCreateOrUpdate(instance)
+	// Get the provisioning interface of the cluster worker nodes
+	if provInterfaceName == "" {
+		r.Log.Info("Provisioning interface name not yet discovered, checking Metal3...")
 
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+		provInterfaceName, err = r.getProvisioningInterface(instance)
 
-	if op != controllerutil.OperationResultNone {
-		r.Log.Info(fmt.Sprintf("Deployment %s successfully reconciled - operation: %s", instance.Name, string(op)))
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if provInterfaceName == "" {
+			return ctrl.Result{}, fmt.Errorf("metal3 provisioning interface configuration not found")
+		}
 	}
 
-	// Get the provisioning IP of node hosting the pod belonging to the CR
-	podIP, err := r.getProvisionServerProvisioningIP(instance)
+	// provisionserver
+	// Create or update the Deployment object
+	op, err := r.deploymentCreateOrUpdate(instance, provInterfaceName)
 
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
-	if podIP == "" {
-		r.Log.Info(fmt.Sprintf("Deployment %s pod provisioning IP not yet available, requeuing and waiting", instance.Name))
-		return ctrl.Result{RequeueAfter: time.Second * 30}, err
+	if op != controllerutil.OperationResultNone {
+		r.Log.Info(fmt.Sprintf("Deployment %s successfully reconciled - operation: %s", instance.Name, string(op)))
 	}
 
-	// Update Status
-	instance.Status.LocalImageURL = r.getLocalImageURL(podIP, instance)
-	err = r.Client.Status().Update(context.TODO(), instance)
-
-	if err != nil {
-		r.Log.Error(err, "Failed to update CR status %v")
-		return ctrl.Result{}, err
+	// Provision IP Discovery Agent sets status' ProvisionIP
+	if instance.Status.ProvisionIP != "" {
+		urlPieces := strings.Split(instance.Status.LocalImageURL, "/")
+		if instance.Status.LocalImageURL == "" || len(urlPieces) <= 1 || (len(urlPieces) >= 3 && urlPieces[2] != instance.Status.LocalImageURL) {
```

Comment:
could you comment here in which cases we update the status, it is a bit hard to follow from this if

Reply:
Sure, my apologies.  I'll clean up the code (hopefully) by using the `net` package, and then add appropriate comments.